crates/sdk/src/build.rs

```diff
@@ -553,52 +573,86 @@ impl AxiomSdk {
-                .file_name(file_name)
-                .mime_str("application/octet-stream")?;
-            form = form.part("config", config_part);
+        // Use a counting reader and perform the request in a background thread while
+        // polling progress from the main thread to update the callback.
+        let uploaded = Arc::new(AtomicU64::new(0));
+        let uploaded_for_thread = Arc::clone(&uploaded);
+        let tar_path_string = tar_path.clone();
+        let url_clone = url.clone();
+        let api_key_owned = self
+            .config
+            .api_key
+            .as_ref()
+            .ok_or_eyre("API key not set")?
+            .to_string();
+        let config_source_for_form = args.config_source.clone();
+
+        let handle = std::thread::spawn(move || -> Result<reqwest::blocking::Response> {
+            let client = Client::builder()
+                .timeout(std::time::Duration::from_secs(300)) // 5 minute timeout
+                .build()?;
+
+            // Open the file and wrap with counting reader
+            let file = File::open(&tar_path_string).context("Failed to open tar file")?;
+            let counting_reader = CountingReader {
+                inner: file,
+                progress: uploaded_for_thread,
+            };
+
+            // Create multipart form
+            let part = reqwest::blocking::multipart::Part::reader(counting_reader)
+                .file_name("program.tar.gz")
+                .mime_str("application/gzip")?;
+
+            let mut form = reqwest::blocking::multipart::Form::new().part("program", part);
+
+            // Add config file if provided
+            if let Some(ConfigSource::ConfigPath(config_path_str)) = config_source_for_form {
+                let config_path = Path::new(&config_path_str);
+                let config_file_content = std::fs::read(config_path).with_context(|| {
+                    format!(
+                        "Failed to read OpenVM config file at: {}",
+                        config_path.display()
+                    )
+                })?;
+                let file_name = config_path
+                    .file_name()
+                    .ok_or_eyre("Invalid config file path")?
+                    .to_string_lossy()
+                    .to_string();
+                let config_part = reqwest::blocking::multipart::Part::bytes(config_file_content)
+                    .file_name(file_name)
+                    .mime_str("application/octet-stream")?;
+                form = form.part("config", config_part);
+            }
+
+            let request = add_cli_version_header(
+                client
+                    .post(url_clone)
+                    .header(API_KEY_HEADER, api_key_owned)
+                    .multipart(form),
+            );
+
+            let response = request.send()?;
+            Ok(response)
+        });
+
+        // Poll progress until the upload thread finishes
```

Comment:
What's the point of spawning thread if we immediately block and wait for completion?